File: sessions/2026-02-04-session25/solutions/cpp/implement_upper_bound_uloamaka.cpp
/* 
    Problem Link: https://www.geeksforgeeks.org/problems/implement-upper-bound/1
*/
class Solution {
  public:
    int upperBound(vector<int>& arr, int target) {
        int left = 0;
        int right = arr.size();
        
        while(left < right){
            int mid = left + (right-left) / 2;

            if (arr[mid] <= target) 
            {
                left = mid+1;
            } 
            else 
            {
                right = mid;
            }
        }
        return left;
    }
};

/* 
    Time complexity : O(log n)
    Space complexity : O(1)
*/

File: sessions/2026-02-04-session25/solutions/go/implement_upper_bound_uloamaka.go
/* 
	Problem Link: https://www.geeksforgeeks.org/problems/implement-upper-bound/1
*/
package main

func upperBound(nums []int, target int) int {
    left, right := 0, len(nums)

    for left < right {
        mid := left + (right-left) / 2

        if nums[mid] <= target {
            left = mid + 1
        } else {
            right = mid
        }
    }

    return left
}

/*
    Time complexity: O(log n)
    Space complexity: O(1)
*/
